cpp/Json.hpp

```diff
@@ -0,0 +1,405 @@
+        Comma, Colon,
+
+        Eof = -1,
+        Error = -2,
+        Empty = -3 // For internal purposes
+    };
+
+    class JsonError : public runtime_error {
+    public:
+        usize position;
+
+        JsonError(const char* msg, usize pos)
+            : runtime_error(msg), position(pos) {}
+    };
+
+    template <typename CharT>
+    class JsonStream {
+        using vstr = basic_string_view<CharT>;
+        using str = basic_string<CharT>;
+
+        std::queue<JsonToken> tokens;
+        std::queue<str> qtok_string;
+        std::queue<i32> qtok_number;
+        JsonToken current_token = JsonToken::Empty;
+        const char* token_kw = nullptr;
+        usize token_position = 0;
+        str token_str;
+
+        usize pos = 0;
+
+    public:
+        std::queue<JsonToken>& get_tokens() {return tokens; }
+        str pop_string() {
+            str s = move(qtok_string.front());
+            qtok_string.pop();
+            return s;
+        }
+        i32 pop_number() {
+            i32 n = qtok_number.front();
+            qtok_number.pop();
+            return n;
+        }
+
+        // If an error occurs, the rest of buffer can be passed in
+        Result<void, JsonError> lex(vstr frag) {
+            static const char* const s_keywords[] = {"null", "true", "false"};
+            for (CharT ch : frag) {
+                ++pos;
+                _reparse:
+                if (current_token == JsonToken::Empty) {
+                    if (std::isspace(ch))
+                        continue;
+                    switch (ch) {
+                        #define CASE_CH(ch, tok) \
+                            case static_cast<CharT>(ch): \
+                                tokens.push(tok); \
+                                break;
+
+                        CASE_CH('[', JsonToken::BArray)
+                        CASE_CH(']', JsonToken::EArray)
+                        CASE_CH('{', JsonToken::BObject)
+                        CASE_CH('}', JsonToken::EObject)
+                        CASE_CH(',', JsonToken::Comma)
+                        CASE_CH('.', JsonToken::Colon)
+                        #undef CASE_CH
+
+                        case static_cast<CharT>('n'):
+                            current_token = JsonToken::Null;
+                            token_kw = s_keywords[0];
+                            ++token_position;
+                            break;
+                        case static_cast<CharT>('t'):
+                            current_token = JsonToken::True;
+                            token_kw = s_keywords[1];
+                            ++token_position;
+                            break;
+                        case static_cast<CharT>('f'):
+                            current_token = JsonToken::False;
+                            token_kw = s_keywords[2];
+                            ++token_position;
+                            break;
+                        case static_cast<CharT>('"'):
```

Comment:
You didn't need to do `'\"'` if I'm right?

Reply:
No, quotes need to be escaped only if surrounded by matching quotes.